cpp/src/groupby/hash/groupby.cu

```diff
@@ -84,42 +91,130 @@ bool constexpr is_hash_aggregation(aggregation::Kind t)
+    // value to be null.
+    column_view null_removed_map(
+      data_type(type_to_id<size_type>()),
+      arg_result->size(),
+      static_cast<void const*>(arg_result->view().template data<size_type>()));
+    auto transformed_result =
+      cudf::detail::gather(table_view({col}),
+                           null_removed_map,
+                           arg_result->nullable() ? cudf::detail::out_of_bounds_policy::IGNORE
+                                                  : cudf::detail::out_of_bounds_policy::NULLIFY,
+                           cudf::detail::negative_index_policy::NOT_ALLOWED,
+                           mr,
+                           stream);
+    return std::move(transformed_result->release()[0]);
+  };
+
+  // Declare overloads for each kind of a agg to dispatch
+  void Dispatch(cudf::detail::min_aggregation& agg) override
+  {
+    if (col.type().id() == type_id::STRING)
+      dense_results->add_result(i, agg, transformed_result(aggregation::ARGMIN));
+    else
+      dense_results->add_result(i, agg, to_dense_agg_result(agg));
+  }
+
+  void Dispatch(cudf::detail::max_aggregation& agg) override
+  {
+    if (col.type().id() == type_id::STRING)
+      dense_results->add_result(i, agg, transformed_result(aggregation::ARGMAX));
+    else
+      dense_results->add_result(i, agg, to_dense_agg_result(agg));
+  }
+
+  void Dispatch(cudf::detail::mean_aggregation& agg) override
+  {
+    if (is_fixed_width(col.type()) and !sparse_results.has_result(i, agg)) {
+      auto sum_agg      = std::make_unique<aggregation>(aggregation::SUM);
+      auto sum_result   = sparse_results.get_result(i, *sum_agg);
+      auto count_agg    = std::make_unique<aggregation>(aggregation::COUNT_VALID);
+      auto count_result = sparse_results.get_result(i, *count_agg);
+      auto result =
+        cudf::detail::binary_operation(sum_result,
+                                       count_result,
+                                       binary_operator::DIV,
+                                       cudf::detail::target_type(col.type(), aggregation::MEAN),
+                                       rmm::mr::get_current_device_resource(),
+                                       stream);
+      sparse_results.add_result(i, agg, std::move(result));
+      dense_results->add_result(i, agg, to_dense_agg_result(agg));
+    }
+  }
+  void Dispatch(cudf::detail::std_var_aggregation& agg) override
+  {
+    // TODO
+  }
+};
+
 // flatten aggs to filter in single pass aggs
 std::tuple<table_view, std::vector<aggregation::Kind>, std::vector<size_t>>
 flatten_single_pass_aggs(std::vector<aggregation_request> const& requests)
 {
   std::vector<column_view> columns;
   std::vector<aggregation::Kind> agg_kinds;
+  std::unordered_set<aggregation::Kind> agg_kinds_set;
   std::vector<size_t> col_ids;
+  auto insert_agg = [&](size_t i, column_view const& request_values, aggregation::Kind k) {
+    if (agg_kinds_set.insert(k).second) {
```

Comment:
This doesn't look like it would work correctly if the same `aggregation::Kind` is requested for multiple columns.

Reply:
fixed.